vshard/replicaset.lua

```diff
@@ -1204,6 +1204,9 @@ local replica_mt = {
         safe_uri = function(replica)
             local uri = luri.parse(replica.uri)
             uri.password = nil
+            if uri.ipv6 then
+                uri.host = '[' .. uri.host .. ']'
+            end
```

Comment:
While it works on the current Tarantool versions, it will not work once `uri.format` is fixed in the core. Your code will then produce incorrect URIs again - with `[[` and `]]`. Fixing it in vshard will not help if we will release at least one vshard version with the current solution - it simply won't work on new Tarantool cores.

The only solution I see is to make it work both for current and future behaviour together. For that you need to do what @Serpentian said: define a helper in `util` module - `uri_format`. It should add the brackets or not, depending on how the core uri module behaves.

A possible solution could be this (in `util.lua`):
```Lua
local uri_v6_str = "[::1]:80"
local uri_format
-- URI module format() function doesn't behave correctly for IPv6 addresses on
-- some Tarantool versions (all <= 3.0). In fact, at the time of writing none of
-- the versions have this bug fixed.
if uri.format(uri.parse(uri_v6_str)) ~= uri_v6_str then
    uri_format = function(u)
        -- Implement here your hack with the brackets.
    end
else
    uri_format = uri.format
end
-- ...
-- In the end of the `util.lua` file add `uri_format` function to the module exports.
```
And then make all the vshard code use `util.uri_format` instead of core's `uri.format`.

Reply:
Fixed